### PR TITLE
Feature atl 5535 implement common datagrid with Multiselect, Sorting and basic datagrid

### PR DESCRIPTION
**Pull request by krutisaklecha**
Added functionalities for new data grid with Basic DataGrid, multiselect and Sorting features.

image
image

### CODE REVIEW

**Review by akate1 — CHANGES_REQUESTED** [review]
Please remove package-lock.json file from PR. We are using only yarn commands on this repo

File: src/components/dataGridInfiniteScroll/dataGridInfiniteScroll.tsx [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+                    );
+                })}
+
+                </tbody>
+            </table>
+        </div>
+
+    )
+}
+export default DataGridInfiniteScroll;
```
[the remark this answers is not in the document]
Comment (krutisaklecha) [reply src/components/dataGridInfiniteScroll/dataGridInfiniteScroll.tsx]: Done

[the remark this answers is not in the document]
Comment (krutisaklecha) [reply src/components/dataGridInfiniteScroll/dataGridInfiniteScroll.tsx]: Done

**Comment by krutisaklecha** [comment]
> Please remove package-lock.json file from PR. We are using only yarn commands on this repo

Done

File: src/components/dataGridInfiniteScroll/datagridInfiniteScrollData.tsx [hunk 2]
```diff
@@ -0,0 +1,1518 @@
+export const DATA = {
```
[the remark this answers is not in the document]
Comment (krutisaklecha) [reply src/components/dataGridInfiniteScroll/datagridInfiniteScrollData.tsx]: Done

File: src/components/dataGridInfiniteScroll/datagridCustomStyles.scss [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    .table {
+
+    border-color: #E4E4E4;
+    border: 1px solid #E4E4E4 !important;
+
+    }
+    table {
+        border: solid 1px #E4E4E4;
+    }
+}
```
[the remark this answers is not in the document]
Comment (krutisaklecha) [reply src/components/dataGridInfiniteScroll/datagridCustomStyles.scss]: Done

File: src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        },
+        useFilters, // Adding the useFilters Hook to the table
+        useSortBy
+    );
+    const [allValues, setIsChecked] = useState<any>([]);
+    useEffect(() => {
+        setIsChecked(rows);
+    }, [rows]);
+
+    const handleChangeSorting = (e) => {
```
Comment (akate1) [inline src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx:109]:
```suggestion
    const handleChangeSorting = (event) => {
```

**Review by akate1 — CHANGES_REQUESTED** [review]

File: src/components/DataGridWithInfiniteScroll/DataGridInfiniteScroll.stories.tsx [hunk 5]
```diff
@@ -11,6 +12,14 @@ const columns = [{accessor: "ip", Header: "IP"}, {accessor: "serial", Header: "S
     Header: "Role"
 },];
 
+const detailColumns = [{accessor: "status", Header: "Status"}, {accessor: "name", Header: "Disk"}, {
```
Comment (akate1) [inline src/components/DataGridWithInfiniteScroll/DataGridInfiniteScroll.stories.tsx:15]: Please move this data in 'DataGridWithInfiniteScrollMockData.ts'

File: src/components/DataGridWithInfiniteScroll/DataGridInfiniteScroll.stories.tsx [hunk 6]
```diff
@@ -21,3 +30,12 @@ storiesOf("Data Grid with Infinite Scroll", module)
...
     </div>))
+    .add("Grid with Column Hide/Show", () => (<div>
+        <DataGridWithInfiniteScroll row={DETAILDATA.data} column={detailColumns} columnSelect={true}/>
+    </div>))
+    .add("Grid with Detail Panel", () => (<div>
+        <DataGridWithInfiniteScroll row={DETAILDATA.data} column={detailColumns} detailPanelShow={true}/>
+    </div>))
+    .add("Empty Datagrid", () => (<div>
+        <DataGridWithInfiniteScroll row={[]} column={columns} style={{height: "70vh"}}/>
+    </div>))
```
Comment (akate1) [inline src/components/DataGridWithInfiniteScroll/DataGridInfiniteScroll.stories.tsx:41]: Please add empty line at EOF

File: src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 7]
```diff
@@ -1,6 +1,11 @@
-import React, {useEffect, useState} from "react";
+import React, {useEffect, useRef, useState} from "react";
 import {useFilters, useSortBy, useTable} from "react-table";
+import { Button} from "@dellstorage/clarity-react/forms/button";
+import { Card, CardBlock, CardTitle } from '@dellstorage/clarity-react/cards'
+import { Table } from '@dellstorage/clarity-react/tables'
 import './DatagridCustomStyles.scss';
+import './DatagridCoulumnSelectionStyles.scss';
+import './DatagridDetailPanelStyles.scss';
```
Comment (akate1) [inline src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx:8]: can we keep single SCSS file for all datagrid CSS ? import these CSS files in 'DataGridWithInfiniteScroll.SCSS'

File: src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 8]
```diff
@@ -55,7 +60,10 @@ type DataGridProps = {
     tableType?: "csg" | "isg" | "compact";
     defaultColumnWidth?: number;
     tableHeight?: number;
+    detailPaneContent? : any;
+    detailPanelShow? : boolean;
```
Comment (akate1) [inline src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx:64]: detailPanelShow -> showDetailPanel

File: src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 9]
```diff
@@ -102,10 +145,102 @@ const DataGridWithInfiniteScroll = (props: DataGridProps) => {
         useSortBy
     );
     const [allValues, setIsChecked] = useState<any>([]);
+    
     useEffect(() => {
-        setIsChecked(rows);
+        if(rows.length != 0){
```
Comment (akate1) [inline src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx:150]: Please use !==

File: src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 10]
```diff
@@ -102,10 +145,102 @@ const DataGridWithInfiniteScroll = (props: DataGridProps) => {
...
+            const HideShowColumnsMenuLeft = refParent.current.getClientRects()[0].width - 180;
+            const transformVal =
+            "translateX(" + HideShowColumnsMenuLeft + "px) " + "translateY(" + HideShowColumnsMenuTop + "px)";
+            settransformVal(transformVal);
+  }
+    setOpen(!isOpen);
+  }
+
+  const handleDetailsPanel = (row, i) => {
+    expandData = [
```
Comment (akate1) [inline src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx:167]: Please move this mock data in separate file

File: src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 11]
```diff
@@ -102,10 +145,102 @@ const DataGridWithInfiniteScroll = (props: DataGridProps) => {
...
+      ]
+
+  
+   setShowDetailData(expandData);
+   setShowDetailPanelTitle(data[i].name)
+    setDetailRowIndex(i);
+    setShowDetailsPanel(true);
+}
+
+const getDetailPanel = () => {
```
Comment (akate1) [inline src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx:209]: getDetailPanel -> buildDetailPanel

File: src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 12]
```diff
@@ -121,7 +256,43 @@ const DataGridWithInfiniteScroll = (props: DataGridProps) => {
...
+            
+                    <Button link className="Reset-button-margin-css"  >
+                    <span className="ColumnSelect-header">Reset Settings </span> </Button>
+
+                </div>
+          </div>
+        }
+
+        <div className="table-css" style={props.style} ref={refParent}>
+            <div className={showDetailsPanel ? 'clr-col-12' : 'clr-col-12'}>
```
Comment (akate1) [inline src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx:295]: What is use of `?` operator here? we are using same CSS Class for both condition

File: src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 13]
```diff
@@ -161,14 +332,20 @@ const DataGridWithInfiniteScroll = (props: DataGridProps) => {
...
+                           </th> }
                     </tr>
                 ))}
                 </thead>
-                <tbody {...getTableBodyProps()} className="table-body">
-                {allValues.map((row, i) => {
+
+
+                { data.length != 0  ?  
+                <tbody {...getTableBodyProps()} className={data.length != 0 ? "table-body" : "empty-datagrid"}>
```
Comment (akate1) [inline src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx:344]: Please use !==

File: src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 14]
```diff
@@ -161,14 +332,20 @@ const DataGridWithInfiniteScroll = (props: DataGridProps) => {
...
+                {allValues ? allValues.map((row, i) => {
                     prepareRow(row);
                     return (
-                        <tr {...row.getRowProps()} className={'csg-row'}>
+                        <tr {...row.getRowProps()} onClick={() => detailPanelShow ? handleDetailsPanel(row,i) : ''} className={'csg-row'}>
```
Comment (akate1) [inline src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx:348]: detailPanelShow && handleDetailsPanel(row,i)

File: src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 15]
```diff
@@ -184,13 +361,37 @@ const DataGridWithInfiniteScroll = (props: DataGridProps) => {
...
+                : 
+                <tbody className="empty-datagrid">
+                 {/* <tr> */}
+                    <div className="datagrid-placeholder-container ng-star-inserted">
+                        <div className="datagrid-placeholder datagrid-empty clr-align-items-center clr-justify-content-center">
+                            <div className="datagrid-placeholder-image ng-star-inserted"></div>
+                            No items found!
+                        </div>
+                    </div>
+                {/* // </tr>  */}
```
Comment (akate1) [inline src/components/DataGridWithInfiniteScroll/DataGridWithInfiniteScroll.tsx:379]: Please remove commented code

File: src/components/DataGridWithInfiniteScroll/DatagridCoulumnSelectionStyles.scss [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+        width: 9rem;
+        display: flex;
+        flex-direction: column;
+        z-index: 1060;
+    }
+    
+    .clr-popover-content {
+        position: fixed;
+        z-index: 5000;
+    }
```
Comment (akate1) [inline src/components/DataGridWithInfiniteScroll/DatagridCoulumnSelectionStyles.scss:51]: Please add an empty line at EOF.
Please use a prettier formatter.

File: src/components/DataGridWithInfiniteScroll/DatagridInfiniteScrollDetailPanelMockData.tsx [hunk 17]
```diff
@@ -0,0 +1,438 @@
...
+                  "partNumber": "TH092WDWGJHDGKFHHKJH",
+                  "SASAddress": "0x87694853ACDB",
+                  "powerStatus": "Spun Up",
+                  "failurePredicted": "No",
+                  "deviceDescription": "Disk 1 of 120 blockplane integrated",
+                  "capacityAvailable": "200 GiB"
+               
+                }
+          
+
```
Comment (akate1) [inline src/components/DataGridWithInfiniteScroll/DatagridInfiniteScrollDetailPanelMockData.tsx:437]: Please use some logic using for/while loops to generate dummy data

**Review by akate1 — CHANGES_REQUESTED** [review]

File: src/components/dataGridInfiniteScroll/DataGridInfiniteScroll.stories.tsx [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+}, {accessor: "template", Header: "Template"}, {accessor: "networking", Header: "Networking"}, {
+    accessor: "role",
+    Header: "Role"
+},];
+
+storiesOf("Data Grid with Infinite Scroll", module)
+    .add("Basic Grid", () => (<div>
+        <DataGridWithInfiniteScroll row={DATA.rows} column={columns}/>
+    </div>))
+    .add("Grid with Multiselect", () => (<div>
```
[the remark this answers is not in the document]
Comment (krutisaklecha) [reply src/components/dataGridInfiniteScroll/DataGridInfiniteScroll.stories.tsx]: Added story for single select

File: src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 19]
```diff
@@ -0,0 +1,197 @@
+import React, {useEffect, useState} from "react";
+import {useFilters, useSortBy, useTable} from "react-table";
+import './DatagridCustomStyles.scss';
```
[the remark this answers is not in the document]
Comment (krutisaklecha) [reply src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx]: Done

File: src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 20]
```diff
@@ -0,0 +1,197 @@
+import React, {useEffect, useState} from "react";
```
[the remark this answers is not in the document]
Comment (krutisaklecha) [reply src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx]: Done

File: src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 21]
```diff
@@ -0,0 +1,197 @@
...
+    row: { [key: string]: any };
+    column: { [key: string]: any };
+    selectionType?: GridSelectionType;
+    sorting?: boolean;
+    expandable?: boolean;
+    expandComponent?: any;
+    pagination?: boolean;
+    columnSelect?: boolean;
+    infiniteScroll?: boolean;
+    tableType?: "csg" | "isg" | "compact";
```
[the remark this answers is not in the document]
Comment (krutisaklecha) [reply src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx]: tableType was not is use, so removed it

File: src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 22]
```diff
@@ -0,0 +1,197 @@
...
+
+type DataGridProps = {
+    className?: string;
+    style?: any;
+    dataqa?: string;
+    id?: string;
+    row: { [key: string]: any };
+    column: { [key: string]: any };
+    selectionType?: GridSelectionType;
+    sorting?: boolean;
```
[the remark this answers is not in the document]
Comment (krutisaklecha) [reply src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx]: As we discussed in one of our call, sorting will be on all columns.

File: src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 23]
```diff
@@ -0,0 +1,197 @@
...
+    expandable?: boolean;
+    expandComponent?: any;
+    pagination?: boolean;
+    columnSelect?: boolean;
+    infiniteScroll?: boolean;
+    tableType?: "csg" | "isg" | "compact";
+    defaultColumnWidth?: number;
+    tableHeight?: number;
+}
+export type DataGridColumn = {
```
[the remark this answers is not in the document]
Comment (krutisaklecha) [reply src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx]: Done

File: src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 24]
```diff
@@ -0,0 +1,197 @@
...
+    className?: string;
+    style?: any;
+    rowID?: number; // not to take from user
+    isSelected?: boolean;
+    disableRowSelection?: boolean;
+};
+const DataGridWithInfiniteScroll = (props: DataGridProps) => {
+    const data: any = props.row;
+    const columns: any = props.column;
+    const selectionType: any = props.selectionType;
```
[the remark this answers is not in the document]
Comment (krutisaklecha) [reply src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx]: Done

File: src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx [hunk 25]
```diff
@@ -0,0 +1,197 @@
...
+    hideSort?: boolean;
+};
+
+type DataGridProps = {
+    className?: string;
+    style?: any;
+    dataqa?: string;
+    id?: string;
+    row: { [key: string]: any };
+    column: { [key: string]: any };
```
[the remark this answers is not in the document]
Comment (krutisaklecha) [reply src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx]: Done

File: src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx (hunk 4 above)
Comment (krutisaklecha) [reply src/components/dataGridInfiniteScroll/DataGridWithInfiniteScroll.tsx]: Done

File: tsconfig.json [hunk 26]
```diff
@@ -21,7 +21,8 @@
     "resolveJsonModule": true,
     "isolatedModules": true,
     "noEmit": false,
-    "jsx": "react-jsx"
+    "jsx": "react-jsx",
+    "noImplicitAny": false
   },
   "include": ["src"]
 }
```
[the remark this answers is not in the document]
Comment (krutisaklecha) [reply tsconfig.json:28]: Done

File: src/components/dataGridInfiniteScroll/DatagridInfiniteScrollMockData.tsx [hunk 27]
```diff
@@ -0,0 +1,1518 @@
...
+        {
+            ip: '192.168.0.2',
+            serial: '678ECG88TGS',
+            model: 'PI-56789672',
+            template: 'ECS EX400',
+            networking: 'DHCP',
+            role: 'Data, Monitor',
+        },
+    ]
+}
```
[the remark this answers is not in the document]
Comment (krutisaklecha) [reply src/components/dataGridInfiniteScroll/DatagridInfiniteScrollMockData.tsx]: Done

File: src/components/DataGridWithInfiniteScroll/DataGridInfiniteScroll.stories.tsx (hunk 6 above)
Comment (krutisaklecha) [reply src/components/DataGridWithInfiniteScroll/DataGridInfiniteScroll.stories.tsx:52]: Done

File: src/components/DataGridWithInfiniteScroll/DatagridInfiniteScrollDetailPanelMockData.tsx (hunk 17 above)
Comment (krutisaklecha) [reply src/components/DataGridWithInfiniteScroll/DatagridInfiniteScrollDetailPanelMockData.tsx:437]: Done

File: src/components/dataGridInfiniteScroll/DataGridInfiniteScroll.stories.tsx [hunk 28]
```diff
@@ -0,0 +1,23 @@
+import {storiesOf} from "@storybook/react";
```
[the remark this answers is not in the document]
Comment (krutisaklecha) [reply src/components/dataGridInfiniteScroll/DataGridInfiniteScroll.stories.tsx]: Done

File: src/components/dataGridInfiniteScroll/DatagridInfiniteScrollMockData.tsx [hunk 29]
```diff
@@ -0,0 +1,1518 @@
...
+            role: 'Data, Monitor',
+        },
+        {
+            ip: '192.168.0.1',
+            serial: '1234ECG88TGS',
+            model: 'PI-12345672',
+            template: 'ECS EX400',
+            networking: 'DHCP',
+            role: 'Data, Monitor',
+        },
```
[the remark this answers is not in the document]
Comment (krutisaklecha) [reply src/components/dataGridInfiniteScroll/DatagridInfiniteScrollMockData.tsx]: Done

File: src/components/DataGridWithInfiniteScroll/DatagridCoulumnSelectionStyles.scss (hunk 16 above)
Comment (krutisaklecha) [reply src/components/DataGridWithInfiniteScroll/DatagridCoulumnSelectionStyles.scss:51]: Done

**Pull request closed without merging**